Load feature flags even in CI

NSeydoux opened the pull request:
Currently, feature flags are correctly loaded locally, but they fail to load in CI. This is due to the fact that they only get loaded after `dotenv` runs, which only happens on dev machines but not in CI.

NSeydoux committed:
Load feature flags even in CI

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -87,36 +87,33 @@ export interface EnvVariables {
 }
 
 export function setupEnv() {
```
I'd rather keep `setupEnv` as simple as possible as it's called in a lot of places.. we've always had it return early in CI (this dates back before we had this library)

ThisIsMissEm replied:
Can the bulk of the changes to this method be reverted? Unless of course you'd argue that they give more clarity to the method?

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/113/commits/9f8f10f1b8cf1c0c52b333146501a3521c1406f8

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -87,36 +87,33 @@ export interface EnvVariables {
   // Creating feature flag object if there are feature flags
   Object.keys(process.env)
     .filter((envVar) => envVar.startsWith("E2E_TEST_FEATURE_"))
     .forEach((key) => (featuredFlags[key] = process.env[key]));
```
This code here is actually not related to setting up the environment (by which we mean, loading dotenv if needed)

This code is actually related to getting the testing environment, so would be better moved to `getBaseTestingEnvironment` (which should have `getTestingEnvironment` inlined)

NSeydoux replied:
Should be fixed in 9615f64

NSeydoux committed:
Refactor environment base loading

Inline common environment variable validation and move the feature flag loading along these.

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -87,69 +86,57 @@ export interface EnvVariables {
+    // Load and validate target environment name.
+    const targetEnvName = process.env.E2E_TEST_ENVIRONMENT;
+    if (
+      !availableEnvironment.includes(targetEnvName as AvailableEnvironment)
```
Ha, just realised this was `availableEnvironment` and not `availableEnvironments` (which is technically more correct give it's an Array.

NSeydoux replied:
9bea33819a3d46115b89929d990f21d0f8dd0df7

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -87,69 +86,57 @@ export interface EnvVariables {
+
+    // Load and validate target OpenID Provider.
+    const targetOp = process.env.E2E_TEST_IDP;
+    if (typeof targetOp !== "string") {
```
Should we URL validate here too?

NSeydoux replied:
897597c5b8122c01bc5e690e7cbdddf38d31bbcc

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -87,69 +86,57 @@ export interface EnvVariables {
+
+    // Creating feature flag object if there are feature flags
+    Object.keys(process.env)
+    .filter((envVar) => envVar.startsWith("E2E_TEST_FEATURE_"))
+    .forEach((key) => (featuredFlags[key] = process.env[key]));
```
I think prettier should suggest this here:

```suggestion
      .filter((envVar) => envVar.startsWith("E2E_TEST_FEATURE_"))
      .forEach((key) => (featuredFlags[key] = process.env[key]));
```

But this still isn't right, it should be:

```suggestion
      .filter((envVar) => envVar.startsWith("E2E_TEST_FEATURE_"))
      .forEach((key) => (featuredFlags[key.substring(17).toLowerCase()] = process.env[key]));
```

So `E2E_TEST_FEATURE_FOO_BAR` would become `features.foo_bar`

Also worth double checking if `featuredFlags` needs to be global still with this change, I don't think it does, I think it can be a `const` inside this method, so a `reduce` instead of `forEach`.

Or:

```typescript
      const featureFlags = Object.keys(process.env)
        .filter((envVar) => envVar.startsWith("E2E_TEST_FEATURE_"))
        .reduce((flags, envVar) => {
          const flag = envVar.substring(17).toLowerCase();
          flags[flag] = process.env[envVar];
          return flags;
        }, {});
```

NSeydoux replied:
Good point, I wasn't super happy about the global variable so here goes: 9f8f10f1b8cf1c0c52b333146501a3521c1406f8, and c840967385486184a502ef832ab0be6f6dcc3bfe

NSeydoux committed:
Return early on env loading

NSeydoux committed:
Manually run prettier

NSeydoux committed:
Remove featureFlags from global scope

NSeydoux committed:
Remove unnecessary todo

NSeydoux committed:
Rename available environments list

NSeydoux committed:
Validate OP is an IRI

NSeydoux committed:
Clear env var prefix

NSeydoux commented on packages/internal-test-env/index.ts:
```diff
@@ -87,69 +86,57 @@ export interface EnvVariables {
   ): T extends NodeVariables ? TestingEnvironmentNode : TestingEnvironmentBrowser   {
     setupEnv();
-    getTestingEnvironment(process.env);
+    // TODO: Replace these inline validations and checks with envalid or env-var
```
b92ed6ee13dc382adc0065f29fdf24df29260637

NSeydoux committed:
Update README

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,97 @@ export interface TestingEnvironmentBase {
-      `We didn't find the given environment variable E2E_TEST_ENVIRONMENT,
-tried looking in the following directory for \`.env.local \`: 
-${envPath}`
+      `We didn't find the given environment variable E2E_TEST_ENVIRONMENT, tried looking in the following directory for '.env.local': ${envPath}`
```
is this good with our line length via prettier?

NSeydoux replied:
I ran prettier manually to check and it didn't complain

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,97 @@ export interface TestingEnvironmentBase {
-    )
-  ) {
+  // Load and validate target OpenID Provider.
+  const targetOp = process.env.E2E_TEST_IDP;
+  if (typeof targetOp !== "string" || !isValidUrl(targetOp)) {
```
```suggestion
  const targetIdp = process.env.E2E_TEST_IDP;
  if (typeof targetIdp !== "string" || !isValidUrl(targetIdp)) {
```

I think that might be clearer?

NSeydoux replied:
OP is a standard OpenID terminology (for OpenID Provider), and I'd prefer we use spec terminology where we can. Changing the env var name has repercussions across all of our repos, so it's not worth the effort though.

ThisIsMissEm replied:
Okay, that's fine then, but let's add support for `E2E_TEST_OP` (though I've only ever heard people inside the company talk of this as IdP), and then we can do a PR to e2e-environments and roll out `E2E_TEST_OP` to all environments and then when we upgrade the developer tools in each SDK, we can switch the CI pipelines across to the new environment variable.

But if we're saying we don't want to migrate to `E2E_TEST_OP`, then I'd rather that this code use `targetIdp` and not `targetOp` for clarity & consistency.

NSeydoux replied:
OP is what the OpenID spec uses, and how everyone in the auth community calls it. We should align on calling it so everywhere, but I'll create a ticket to do it separately not to expand the scope of this specific PR.

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,97 @@ export interface TestingEnvironmentBase {
-      `Unknown environment: [${
-        (environment as EnvVariables).E2E_TEST_ENVIRONMENT
-      }]`
+      `The environment variable E2E_TEST_IDP is not a valid URL: found ${targetOp}.`
```
```suggestion
      `The environment variable E2E_TEST_IDP is not a valid URL: found ${targetIdp}.`
```

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/113/commits/12ccd9e975de80a287fbca4f76c622870c49ebd6

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,97 @@ export interface TestingEnvironmentBase {
+    .reduce(
+      (featureFlags, envVar) => ({
+        ...featureFlags,
+        [`${envVar}`]: process.env[envVar]?.split(ENV_VAR_PREFIX)[1],
```
I don't think this is correct? See my previous comment — the feature flag name would need to be `envVar.substring(ENV_VAR_PREFIX.length)` or `envVar.split(ENV_VAR_PREFIX)[1]` (I prefer the substring version)

This code would be splitting the value of the environment variable.

NSeydoux replied:
Oh yeah you're right this is backwards ><. And you're right, I'll use substrings, it is a bit clearer.

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/113/commits/559bfd0e3202f0cc2b77abb0fa8691fbd5ee9de6

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,97 @@ export interface TestingEnvironmentBase {
-  
-    return libVars ? merge(base, validateLibVars(libVars)) : base;
+  const base = {
+    idp: targetOp,
```
```suggestion
    idp: targetIdp,
```

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/113/commits/12ccd9e975de80a287fbca4f76c622870c49ebd6

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,97 @@ export interface TestingEnvironmentBase {
-    // Enforce client credentials are present for the resource owner.
-    clientCredentials: { owner: { id: true, secret: true }}
-  }) as NodeVariables);
+  return getBaseTestingEnvironment(
```
Perhaps `getBaseTestingEnvironment<NodeVariables>(` instead? That way you don't need the `as NodeVariables` afaict.

NSeydoux replied:
https://github.com/inrupt/typescript-sdk-tools/pull/113/commits/5493567377bb3b9cf8b2652473911fa2329339f8

ThisIsMissEm reviewed (changes requested):
I don't think the feature flag construction is correct still.

NSeydoux committed:
Fix feature flags capture

NSeydoux committed:
Change targetOp to targetIdp

Let's align on the latter in a future commit

NSeydoux committed:
Use generics instead of cast

NSeydoux committed:
Format

ThisIsMissEm commented on packages/internal-test-env/README.md:
```diff
@@ -13,3 +13,30 @@ Add the function import as needed.
+These two functions will load from the environment variables as follows:
+- Some variables will always be loaded and validated if applicable 
+  - `E2E_TEST_ENVIRONMENT`: name of the target environment, used for information purpose.
+    E.g., `Dev-Next`
```
```suggestion
    E.g., `ESS Dev-Next`
```

Looks like the string was incorrect here

ThisIsMissEm commented on packages/internal-test-env/index.ts:
```diff
@@ -78,100 +77,96 @@ export interface TestingEnvironmentBase {
 type FeatureFlags = {
   [key: string]: any;
 };
+const ENV_VAR_PREFIX = "E2E_TEST_FEATURE_";
```
This might be clearer as, though change isn't needed
```suggestion
const ENV_FEATURE_PREFIX = "E2E_TEST_FEATURE_";
```

NSeydoux replied:
ENV_FEATURE_PREFIX

ThisIsMissEm reviewed (approved):
Looks good to me, just ~three~ two small comments optional to address

NSeydoux committed:
Update packages/internal-test-env/README.md

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

NSeydoux committed:
Rename constant
